File: Day01FSharp/Day01.fs
open System
open System.IO

let part1 (numbers: int list): int =
    let sorted = List.sort numbers
    let set = Set.ofSeq numbers
    let n = List.find (fun x -> set.Contains (2020 - x)) sorted
    n * (2020 - n)

let part2 (numbers: int list): int =
    let set = Set.ofSeq numbers
    let indexed = List.indexed numbers
    let pairs =
        Seq.allPairs indexed indexed
        |> Seq.filter (fun (ia, ib) -> (fst ia) < (fst ib))
        |> Seq.map (fun (ia, ib) -> (snd ia), (snd ib))
    let (a, b) = Seq.find (fun (a, b) -> set.Contains (2020 - a - b)) pairs
    a * b * (2020 - a - b)

[<EntryPoint>]
let main argv =
    let numbers = File.ReadAllLines "day01.txt" |> Seq.map Int32.Parse |> List.ofSeq
    printfn "Part 1: %d" (part1 numbers)
    printfn "Part 2: %d" (part2 numbers)
    0

File: Day01/Day01.cs
using System;
using System.IO;
using System.Linq;

var lines = File.ReadLines("day01.txt");
var numbers = lines.Select(int.Parse).OrderBy(x => x).ToList();

void Part1()
{
	for (int i = 0; i < numbers.Count - 1; i++)
	{
		for (int j = numbers.Count - 1; j > i; j--)
		{
			if (numbers[i] + numbers[j] == 2020)
			{
				Console.WriteLine($"Part 1: {numbers[i] * numbers[j]}");
				return;
			}
			else if (numbers[i] + numbers[j] < 2020)
			{
				break;
			}
		}
	}
}

void Part2()
{
	for (int i = 0; i < numbers.Count - 2; i++)
	{
		for (int j = i + 1; j < numbers.Count - 1; j++)
		{
			for (int k = numbers.Count - 1; k > j; k--)
			{
				if (numbers[i] + numbers[j] + numbers[k] == 2020)
				{
					Console.WriteLine($"Part 2: {numbers[i] * numbers[j] * numbers[k]}");
					return;
				}
				else if (numbers[i] + numbers[j] + numbers[k] < 2020)
				{
					break;
				}
			}
		}
	}
}

Part1();
Part2();
